AMBARI-24986. Using Ambari server's CLI to enable/disable Trusted Proxy support

smolnar82 opened the pull request:
## What changes were proposed in this pull request?

A brand new option has been added to our `ambari-server` command line tool to setup Trusted Proxy support: `ambari-server setup-tproxy`

Using this option the end user will have the ability to configure the following:
- either enable Trusted Proxy support with an arbitrary number of proxied users (hosts, users and groups can be specified)
- or the end user can disable an already configured Trusted Proxy setup

 
## How was this patch tested?

Implemented a new unit test class to the the entire functionality:
```
$ mvn -Del.log=WARN -Dcheckstyle.skip -Dfindbugs.skip -Drat.skip -Dswagger.skip -DskipSurefireTests=true -Dpython.test.mask=TestSetupTrustedProxy.py clean test
...
test_disabling_tproxy_support (TestSetupTrustedProxy.TestSetupTrustedProxy) ... ok
test_invalid_tproxy_enabled_cli_option_should_result_in_error (TestSetupTrustedProxy.TestSetupTrustedProxy) ... ok
test_silent_mode_is_not_allowed (TestSetupTrustedProxy.TestSetupTrustedProxy) ... ok
test_tproxy_is_enabled_for_two_proxy_users (TestSetupTrustedProxy.TestSetupTrustedProxy) ... ok
test_tproxy_setup_should_fail_if_server_is_not_running (TestSetupTrustedProxy.TestSetupTrustedProxy) ... ok

----------------------------------------------------------------------
Ran 5 tests in 0.005s

OK
----------------------------------------------------------------------
Total run:5
Total errors:0
Total failures:0
OK
[INFO] 
[INFO] --- maven-checkstyle-plugin:2.17:check (checkstyle) @ ambari-server ---
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time: 43.508 s
[INFO] Finished at: 2018-12-03T16:06:00+01:00
[INFO] Final Memory: 105M/1779M
[INFO] ------------------------------------------------------------------------
```

In addition to uni testing I tested the tool manually using different setups and CLI options. For example:

1. Enabling Trusted Proxy for `knox` and `admin`:
```
$ ambari-server setup-tproxy --ambari-admin-username=admin --ambari-admin-password=admin -v
Using python  /usr/bin/python
Setting up Trusted Proxy support...
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Setup Trusted Proxy
INFO: about to run command: ps -p 19757
INFO: 
process_pid=4793
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Fetching Trusted Proxy configuration from DB
INFO: Fetching information from Ambari's REST API
INFO: 
Trusted Proxy support is currently not configured

Do you want to configure Trusted Proxy Support [y/n] (y)? y
The proxy user's (local) username? knox
Allowed hosts for knox (*)? knox_hosts
Allowed users for knox (*)? knox_users
Allowed groups for knox (*)? knox_groups
Add another proxy user [y/n] (n)? y
The proxy user's (local) username? admin
Allowed hosts for admin (*)? admin_hosts
Allowed users for admin (*)? admin_users
Allowed groups for admin (*)? admin_groups
Add another proxy user [y/n] (n)? n
Ambari Server 'setup-tproxy' completed successfully.
```

Fetching `tproxy-configurations`:
<img width="902" alt="screen shot 2018-12-03 at 4 10 51 pm" src="https://user-images.githubusercontent.com/34065904/49382649-3af14900-f717-11e8-89ee-b4fa8e649fca.png">


2. Disabling Trusted Proxy support:
```
$ ambari-server setup-tproxy --ambari-admin-username=admin --ambari-admin-password=admin -v
Using python  /usr/bin/python
Setting up Trusted Proxy support...
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Setup Trusted Proxy
INFO: about to run command: ps -p 19757
INFO: 
process_pid=4950
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Fetching Trusted Proxy configuration from DB
INFO: Fetching information from Ambari's REST API
INFO: Received HTTP 200 while fetching information from Ambari's REST API
INFO: 
Trusted Proxy support is currently enabled

Do you want to disable Trusted Proxy support [y/n] (n)? y
Ambari Server 'setup-tproxy' completed successfully.
```

Fetching `tproxy-configurations`:
<img width="1472" alt="screen shot 2018-12-03 at 4 11 28 pm" src="https://user-images.githubusercontent.com/34065904/49382681-552b2700-f717-11e8-8b76-6d5bc7124487.png">

smolnar82 committed:
AMBARI-24986. Using Ambari server's CLI to enable/disable Trusted Proxy support

rlevas commented on ambari-server/src/main/python/ambari-server.py:
```diff
@@ -600,11 +601,20 @@ def init_setup_sso_options(parser):
+  parser.add_option('--ambari-admin-username', default=None, help="Ambari administrator username for accessing Ambari's REST API", dest="ambari_admin_username")
+  parser.add_option('--ambari-admin-password', default=None, help="Ambari administrator password for accessing Ambari's REST API", dest="ambari_admin_password")
+  parser.add_option('--tproxy-enabled', default=None, help="Indicates whether to enable/disable Trusted Proxy Support", dest="tproxy_enabled")
+
```
Maybe allow for one proxy user to be set up via the CLI args?
- `--tproxy-proxy-user`
- `--tproxy-proxyuser-hosts`
- `--tproxy-proxyuser-users`
- `--tproxy-proxyuser-groups`

Example:
```
ambari-server setup-trusted-proxy -ambari-admin-username=admin --tproxy-enabled=true --tproxy-proxy-user=knox --tproxy-proxyuser-hosts="c7401.ambari.apache.org" --tproxy-proxyuser-users="*" --tproxy-proxyuser-users="users"
```

smolnar82 replied:
I'm OK with this; but we may go further. What if we supported the following:

```
--tproxy-PROXYUSER-hosts
--tproxy-PROXYUSER-users
--tproxy-PROXYUSER-groups
```

And we would check if there is at least any of `proxy-X-[hosts|users|groups]` is defined, then we should have the rest (i.e `--proxy-X-hosts` and `proxy-X-groups` are defined but there is no `proxy-X-users` then we fail the operation).

Moreover: if there is at least one full configuration of a proxyuser set using these options there should be no more proxy users asked in interactively

This way we allow end-users to define any number of proxy user configurations without user interaction (thus this tool would be 100% scriptable).

rlevas replied:
@smolnar82 , that works for me. Your way is more flexible. :)

smolnar82 replied:
@rlevas 

As discussed over Zoom the above recommended approach does not really work due to CLI option limitations.  Instead the following CLI option will be defined: `--tproxy-configuration-file-path`; if this is provided a file must exists on the referenced path and must contain the Trusted Proxy configurations in JSON format.

For instance:
```
--tproxy-configuration-file-path=/my/path/tproxy.configs

$ cat /my/path/tproxy.configs
[
  {
    "name"   : "knox",
    "hosts"  : "host1",
    "users"  : "user1",
    "groups" : "group1"
  },
  {
    "name"   : "admin",
    "hosts"  : "host2",
    "users"  : "user2, user3",
    "groups" : "group2"
  }
]
```

rlevas replied:
Nice... but maybe change `name` to `proxyuser`?

smolnar82 replied:
Sure

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4677/
Test PASSed.

rlevas commented on ambari-server/sbin/ambari-server:
```diff
@@ -188,6 +188,10 @@ case "${1:-}" in
         echo -e "Setting up SSO authentication properties..."
         $PYTHON "$AMBARI_PYTHON_EXECUTABLE" "$@"
         ;;
+  setup-tproxy)
```
Change `setup-tproxy` to `setup-trusted-proxy`

smolnar82 replied:
ok

rlevas commented on ambari-server/sbin/ambari-server:
```diff
@@ -210,7 +214,7 @@ case "${1:-}" in
   *)
         echo "Usage: $AMBARI_EXECUTABLE
-        {start|stop|reset|restart|upgrade|status|upgradestack|setup|setup-jce|setup-ldap|sync-ldap|set-current|setup-security|refresh-stack-hash|backup|restore|update-host-names|check-database|enable-stack|setup-sso|db-purge-history|install-mpack|uninstall-mpack|upgrade-mpack|setup-kerberos|setup-pam|migrate-ldap-pam} [options]
+        {start|stop|reset|restart|upgrade|status|upgradestack|setup|setup-jce|setup-ldap|sync-ldap|set-current|setup-security|refresh-stack-hash|backup|restore|update-host-names|check-database|enable-stack|setup-sso|setup-tproxy|db-purge-history|install-mpack|uninstall-mpack|upgrade-mpack|setup-kerberos|setup-pam|migrate-ldap-pam} [options]
```
Change `setup-tproxy` to `setup-trusted-proxy`

rlevas commented on ambari-server/src/main/python/ambari_server/setupActions.py:
```diff
@@ -50,3 +50,4 @@
 PAM_SETUP_ACTION = "setup-pam"
 MIGRATE_LDAP_PAM_ACTION = "migrate-ldap-pam"
 KERBEROS_SETUP_ACTION = "setup-kerberos"
+SETUP_TPROXY_ACTION = "setup-tproxy"
```
Change `setup-tproxy` to `setup-trusted-proxy`

rlevas commented on ambari-server/src/main/python/ambari_server/setupTrustedProxy.py:
```diff
@@ -0,0 +1,154 @@
+      remove_tproxy_conf(ambari_properties, admin_login, admin_password)
+
+  else:
+    warning = "setup-tproxy is not enabled in silent mode."
```
Change `setup-tproxy` to `setup-trusted-proxy`

rlevas commented on ambari-server/src/test/python/TestSetupTrustedProxy.py:
```diff
@@ -0,0 +1,230 @@
+      setup_trusted_proxy(options)
+      self.fail("Should fail with fatal exception")
+    except NonFatalException as e:
+      self.assertTrue("setup-tproxy is not enabled in silent mode." in e.reason)
```
Change `setup-tproxy` to `setup-trusted-proxy`

zeroflag commented on ambari-server/src/main/python/ambari_server/setupTrustedProxy.py:
```diff
@@ -0,0 +1,154 @@
+    else:
+      raise http_error
+
+  if json_data and 'Configuration' in json_data and 'properties' in json_data['Configuration']:
```
Consider using a_dict.get(key, default) instead of the _in_ checks 

For example: 

```python
json_data.get('Configuration', {}).get('properties', {})
```

zeroflag commented on ambari-server/src/main/python/ambari_server/setupTrustedProxy.py:
```diff
@@ -0,0 +1,154 @@
+from ambari_server.userInput import get_validated_string_input, get_YN_input
+
+
+PROXYUSER_PLACEHOLDER = "$PROXY_USER"
```
If the placeholder should be always replaced in the end I would use format strings (either with %s or {some_name} + .format(..) ) here.

zeroflag commented on ambari-server/src/main/python/ambari_server/setupTrustedProxy.py:
```diff
@@ -0,0 +1,154 @@
+      tproxy_support_enabled = get_value_from_dictionary(properties, TPROXY_SUPPORT_ENABLED)
+
+      if tproxy_support_enabled:
+        if tproxy_support_enabled and 'true' == tproxy_support_enabled:
```
why do we need to check the tproxy_support_enabled once again here?

smolnar82 replied:
At this stage `tproxy_support_enabled` is a String and can be undefined.

zeroflag replied:
But we're checking in line 123 and in line 124 too.

```python
  if tproxy_support_enabled:
    if tproxy_support_enabled and ..:
```

dlysnichenko commented on ambari-server/src/main/python/ambari_server/setupTrustedProxy.py:
```diff
@@ -0,0 +1,154 @@
+  try:
+    response_code, json_data = get_json_via_rest_api(ambari_properties, admin_login, admin_password, TPROXY_CONFIG_API_ENTRYPOINT)
+  except urllib2.HTTPError as http_error:
+    if http_error.code == 404:
```
how about using named constants? 
See that table after `and also the following constants for integer status codes:` at https://docs.python.org/2/library/httplib.html

smolnar82 replied:
Thanks, I was not aware of that module; will replace it

smolnar82 committed:
AMBARI-24986. Code improvements based on review comments

smolnar82 commented:
@rlevas , @zeroflag , @dlysnichenko 

Thank you all for your comments; I incorporated them and added a new CLI option to support Trusted Proxy setup in non-interactive mode (see `--tproxy-configuration-file-path`).

In addition to uni testing I executed some E2E tests - as usual - to make sure the tool works as expected with the following outcome:
```
$ ambari-server setup-trusted-proxy  --ambari-admin-username=admin --ambari-admin-password=admin --tproxy-enabled=true --tproxy-configuration-file-path=my/path/tproxy.configs -v
Using python  /usr/bin/python
Setting up Trusted Proxy support...
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Setup Trusted Proxy
INFO: about to run command: ps -p 19757
INFO: 
process_pid=15668
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Fetching Trusted Proxy configuration from DB
INFO: Fetching information from Ambari's REST API
Ambari Server 'setup-trusted-proxy' completed successfully.

$ cat my/path/tproxy.configs 
[
  {
    "proxyuser"   : "knox",
    "hosts"  : "host1",
    "users"  : "user1",
    "groups" : "group1"
  },
  {
    "proxyuser"   : "admin",
    "hosts"  : "host2",
    "users"  : "user2, user3",
    "groups" : "group2"
  }
]

$ambari-server setup-trusted-proxy  --ambari-admin-username=admin --ambari-admin-password=admin --tproxy-enabled=true --tproxy-configuration-file-path=my/path/nonExisting.configs -v
Using python  /usr/bin/python
Setting up Trusted Proxy support...
INFO: Loading properties from /etc/ambari-server/conf/ambari.properties
INFO: Setup Trusted Proxy
INFO: about to run command: ps -p 19757
INFO: 
process_pid=16144
ERROR: Exiting with exit code 1. 
REASON: The following errors occurred while processing your request: ['--tproxy-configuration-file-path is set to a non-existing file: my/path/nonExisting.configs']
```

<img width="950" alt="screen shot 2018-12-04 at 9 50 17 pm" src="https://user-images.githubusercontent.com/34065904/49472006-a9fd9900-f80e-11e8-8a2a-03f84ee2df96.png">

rlevas commented on ambari-server/src/main/python/ambari-server.py:
```diff
@@ -600,11 +602,25 @@ def init_setup_sso_options(parser):
+  parser.add_option('--tproxy-enabled', default=None, help="Indicates whether to enable/disable Trusted Proxy Support", dest="tproxy_enabled")
+  parser.add_option('--tproxy-configuration-file-path', default=None,
+                    help="The path where the Trusted Proxy configuration is located. The content is expected to be in JSON format." \
+                    "Sample configuration:[{\"name\": \"knox\", \"hosts\": \"host1\", \"users\": \"user1, user2\", \"groups\": \"group1\"}]",
```
Sample configuration reads "name" rather than "proxyuser".

smolnar82 replied:
Done

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4685/
Test PASSed.

asfgit commented:
Refer to this link for build results (access rights to CI server needed): 
https://builds.apache.org/job/Ambari-Github-PullRequest-Builder/4686/
Test PASSed.
